### PR TITLE
Underline error listener

### PR DESCRIPTION
**Pull request by laurits-mumberg**

### CODE REVIEW

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        String input = tokens.getTokenSource().getInputStream().toString();
+        String[] lines = input.split("\n");
+        String errorLine = lines[line - 1];
+
+        //Gets the token that caused the error. Gets parsed from Object to token to get position
+        Token offendingToken = (Token)offendingSymbol;
+        Interval offendingTokenInterval = new Interval(offendingToken.getStartIndex(), offendingToken.getStopIndex());
+
+        ThrowUnderlinedError(msg, offendingToken);
+    }
```
Comment (humleflue) [inline ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: Maybe we should abstract this away, if we aren't supposed to use it. 
We could make a super class?

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        for (Token offendingToken : offendingTokens) {
+            int lineNumber = offendingToken.getLine();
+
+            //Checks whether or not an existing offending token is present in a line.
+            LineWithError errorLine = null;
+            for (LineWithError line : errorLines){
+                if (line.lineNumber == lineNumber){
+                    errorLine = line;
+                }
+            }
```
Comment (humleflue) [inline ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java:109]: I would like to see this abstracted away into an auxiliary method.

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+     * @param offendingTokenInterval
+     */
+    private void printUnderlinedIntervals(ArrayList<Interval> offendingTokenInterval) {
+        int printPos = 0;
+
+        for (Interval interval : offendingTokenInterval){
+            while (printPos < interval.a) {
+                System.out.print(" ");
+                printPos++;
+            }
```
Comment (humleflue) [inline ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: I would like to see this abstracted away into an auxiliary method. Also, I'd like to see print only being called once.

Comment (humleflue) [reply ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: Or maybe even better - just don't loop at all.
```suggestion
            if (printPos < interval.a) {
                System.out.print(" ".repeat(interval.a - printPos));
                printPos = interval.a;
            }
```

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            if ( interval.a >= 0 && interval.b >= 0) {
+                for (int i = interval.a; i <= interval.b; i++) {
+                    printRed("^");
+                    printPos++;
+                }
```
Comment (humleflue) [inline ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: This and the above requires some testing. But something like this should work:
```suggestion
                printRed("^".repeat(interval.b - interval.a));
                printPos += interval.b - interval.a;
```

File: ANTLR/src/Compiler/Main.java [hunk 5]
```diff
@@ -22,33 +23,47 @@ public static void main(String[] args) throws IOException {
...
-        OutputFile output = new OutputFile(targetCode, args);
-        output.execute();
+            // Code generation
+            JavaScriptCodeGenerationVisitor codeGenerator = new JavaScriptCodeGenerationVisitor();
+            String targetCode = codeGenerator.visit(tree);
+
+            OutputFile output = new OutputFile(targetCode, args);
+            output.execute();
+        }catch (Exception e){
+            //Do not know what makes sense to do here if anything, so empty for now.
```
Comment (humleflue) [inline ANTLR/src/Compiler/Main.java]: When will this happen?

File: ANTLR/src/Compiler/SymbolTable/SymbolTableGeneratorListener.java [hunk 6]
```diff
@@ -2,17 +2,23 @@
...
 
 public class SymbolTableGeneratorListener extends LangBaseListener{
+    private UnderlineErrorListener errorListener;
     public ParseTreeProperty<Scope> scopes = new ParseTreeProperty<>();
     public BaseScope globalScope = new BaseScope();
     Scope currentScope;
 
+    public SymbolTableGeneratorListener(UnderlineErrorListener errorListener) {
+        this.errorListener = errorListener;
+    }
```
Comment (humleflue) [inline ANTLR/src/Compiler/SymbolTable/SymbolTableGeneratorListener.java]: Hvis du alligevel ikke bruger et interface til error listeneren, så ser jeg den hellere blive instansieret inde i klassen.

Comment (humleflue) [inline ANTLR/src/Compiler/SymbolTable/SymbolTableGeneratorListener.java]:
```suggestion
```

File: ANTLR/test/tests/SymbolTable/SymbolTableTests.java [hunk 7]
```diff
@@ -19,7 +20,7 @@
     private SymbolTableGeneratorListener getWalker(ParseTree tree) {
         ParseTreeWalker walker = new ParseTreeWalker();
 
-        SymbolTableGeneratorListener symbolTableGeneratorListener = new SymbolTableGeneratorListener();
+        SymbolTableGeneratorListener symbolTableGeneratorListener = new SymbolTableGeneratorListener(new UnderlineErrorListener());
```
Comment (humleflue) [inline ANTLR/test/tests/SymbolTable/SymbolTableTests.java:23]:
```suggestion
        SymbolTableGeneratorListener symbolTableGeneratorListener = new SymbolTableGeneratorListener();
```

File: ANTLR/src/Compiler/SymbolTable/SymbolTableGeneratorListener.java [hunk 8]
```diff
@@ -2,17 +2,23 @@
...
 import Compiler.AntlrGenerated.LangParser.*;
+import Compiler.ErrorHandling.UnderlineErrorListener;
 import org.antlr.v4.runtime.ParserRuleContext;
 import org.antlr.v4.runtime.tree.ParseTreeProperty;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class SymbolTableGeneratorListener extends LangBaseListener{
+    private UnderlineErrorListener errorListener;
```
Comment (humleflue) [inline ANTLR/src/Compiler/SymbolTable/SymbolTableGeneratorListener.java]:
```suggestion
    private UnderlineErrorListener errorListener = new UnderlineErrorListener();
```

File: ANTLR/test/tests/TypeCheckerVisitor/TypeCheckerTestsBase.java [hunk 9]
```diff
@@ -20,9 +21,10 @@ protected void generateTreeWithSymbols(String testSourceCode) throws IOException
         tree = TestCase.createTree(testSourceCode);
 
         ParseTreeWalker walker = new ParseTreeWalker();
-        symbolTableGeneratorListener = new SymbolTableGeneratorListener();
+
+        symbolTableGeneratorListener = new SymbolTableGeneratorListener(new UnderlineErrorListener());
```
Comment (humleflue) [inline ANTLR/test/tests/TypeCheckerVisitor/TypeCheckerTestsBase.java:25]:
```suggestion
        symbolTableGeneratorListener = new SymbolTableGeneratorListener();
```

**Review by humleflue — CHANGES_REQUESTED** [review]

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java (hunk 1 above)
Comment (laurits-mumberg) [reply ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: it is now interface

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java (hunk 3 above)
Comment (laurits-mumberg) [reply ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: ok

File: ANTLR/src/Compiler/Main.java (hunk 5 above)
Comment (laurits-mumberg) [reply ANTLR/src/Compiler/Main.java]: Whenever ThrowError is called in the errorListener. / Whenever an error is occurred.

File: ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java (hunk 4 above)
Comment (laurits-mumberg) [reply ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java]: smart

**Comment by laurits-mumberg** [comment]
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

**Review by humleflue — APPROVED** [review]
LGTM :trollface:

**Pull request merged**